Update checkpoint vocabulary

anderleich opened the pull request:
Draft PR to add new vocabulary to existing checkpoint by reusing checkpoint's embeddings

This is what the new training procedure would be:

1. Run build_vocab as usual. This would generate the vocabulary files for the new corpora.
2. Then in the training procedure:
- Load checkpoint (fields would be populated with the checkpoint’s vocab)
- Build fields for the new vocabulary as usual from the vocabulary files generated in step1.
- Extend checkpoint fields vocabulary with the new vocabulary fields (only appending new words at the end, as stated in Torch docs)
- Assign the extended vocabulary to the new models fields
- In build_base_model, after loading the checkpoint’s state_dicts, replace new model’s embeddings from 0 to len(checkpoint.vocab_size) with checkpoint’s embeddings. It should be in the same order as in the extended vocabulary new words were appended at the end.
- Remove embeddings parameters from checkpoint as new embeddings have been already added to the model
- Continue as usual

francoishernandez commented on onmt/inputters/inputter.py:
```diff
@@ -288,6 +287,17 @@ def _build_field_vocab(field, counter, size_multiple=1, **kwargs):
         _pad_vocab_to_multiple(field.vocab, size_multiple)
 
 
+def _update_field_vocab(field, old_field):
+    new_vocabulary = []
+    for w in field.vocab.itos:
+        if w not in old_field.vocab.stoi:
+            new_vocabulary.append(w)
+    logger.info('New vocabulary has %d new tokens.' % (len(new_vocabulary)))
+    print(new_vocabulary)
+    old_field.vocab.extend(field.vocab)
+    field.vocab = old_field.vocab
```
What happens here in terms of `freqs` for the new tokens? E.g. if we would like to add a few known tokens to the vocab, without necessarily calling `build_vocab` again, how would that be handled? Should we always have some 'licit' frequency in the new vocab?

anderleich replied:
The main use case I was thinking is finetuning a previously trained model with a new corpus which potentially has new tokens in vocabulary (E.g. new protected tokens for domain tagging). In such cases we need to add these tokens to the vocabulary accepted by the model and thus build vocabulary would be run for the new corpus. 
I don't know exactly what you mean by the `freqs` for the new tokens. As far as I know, token frequencies are just to remove those which are below a given frequency (E.g. `src_vocab_threshold`). Which would be a use case in which it wouldn't be necessary to call build_vocab or have a 'licit' frequency?

francoishernandez commented on onmt/bin/train.py:
```diff
@@ -53,6 +52,33 @@ def prepare_fields_transforms(opt):
+
+    if opt.dump_fields:
+        save_fields(fields, opt.save_data, overwrite=opt.overwrite)
+    if opt.dump_transforms or opt.n_sample != 0:
+        transforms = make_transforms(opt, transforms_cls, fields)
+    if opt.dump_transforms:
+        save_transforms(transforms, opt.save_data, overwrite=opt.overwrite)
+    if opt.n_sample != 0:
+        logger.warning(
+            "`-n_sample` != 0: Training will not be started. "
+            f"Stop after saving {opt.n_sample} samples/corpus.")
+        save_transformed_sample(opt, transforms, n_sample=opt.n_sample)
+        logger.info(
+            "Sample saved, please check it before restart training.")
+        sys.exit()
```
Why duplicating this?

anderleich replied:
It is a copy of `prepare_fields_transforms` which is only run when not checkpoint is given (fresh start). So I saw the need to give those options when updating vocabulary too. Maybe it could be generalized?

francoishernandez commented on onmt/bin/train.py:
```diff
@@ -53,6 +52,33 @@ def prepare_fields_transforms(opt):
+    fields = build_update_dynamic_fields(
+        opt, old_fields, src_specials=specials['src'], tgt_specials=specials['tgt'])
+
+    # maybe prepare pretrained embeddings, if any
+    prepare_pretrained_embeddings(opt, fields)
```
Not compatible with the update, or is it?

anderleich replied:
Not sure. I'll give it thought

francoishernandez reviewed (commented):
A few comments. Not exhaustive.
@Zenglinxiao can you check this out as well?

Zenglinxiao reviewed (commented):
It seems the current `update_vocab` only supports adding new vocabulary to the previous one. What if we want to delete some?
Also, we could not change the order by increasing the frequency of some vocabulary in the current implementation.
For reference, in the OpenNMT-tf, they stored the [mapping](https://github.com/OpenNMT/OpenNMT-tf/blob/bbf99f40ba0839d1a0541debbf0dde920f904a87/opennmt/data/vocab.py#L286) between the new vocab and old vocab which supports both add/delete/reorder.

NOTE: Remember to reuse the common code when you finish construction.
PS: I personally prefer to make this a separate script creating an updated checkpoint rather than including it to `train.py`. What do you think? @francoishernandez

anderleich commented on onmt/opts.py:
```diff
@@ -225,6 +225,8 @@ def model_opts(parser):
     group.add('--position_encoding', '-position_encoding', action='store_true',
               help="Use a sin to mark relative words positions. "
                    "Necessary for non-RNN style models.")
+    group.add("-update_embeddings", "--update_embeddings", action="store_true",
```
Ok

anderleich commented on onmt/train_single.py:
```diff
@@ -31,6 +31,8 @@ def _get_model_opts(opt, checkpoint=None):
             # ensure tensorboard output is written in the directory
             # of previous checkpoints
             opt.tensorboard_log_dir_dated = model_opt.tensorboard_log_dir_dated
+        if opt.update_embeddings:
```
Ok

anderleich commented on onmt/model_builder.py:
```diff
@@ -232,6 +232,19 @@ def fix_key(s):
         # end of patch for backward compatibility
 
+        if model_opt.update_embeddings:
+            enc_emb_name = "encoder.embeddings.make_embedding.emb_luts.0.weight"
```
Ok

anderleich commented on onmt/inputters/inputter.py:
```diff
@@ -366,6 +386,23 @@ def _build_fields_vocab(fields, counters, data_type, share_vocab,
+    _update_fv_from_multifield(tgt_multifield, old_fields["tgt"])
+    src_multifield = fields["src"]
+    _update_fv_from_multifield(src_multifield, old_fields["src"])
+    '''
+    if opts.share_vocab:
+        # `tgt_vocab_size` is ignored when sharing vocabularies
+        logger.info(" * merging src and tgt vocab...")
+        src_field = src_multifield.base_field
+        tgt_field = tgt_multifield.base_field
+        _merge_field_vocabs(
+            src_field, tgt_field)
+        logger.info(" * merged vocab size: %d." % len(src_field.vocab))
+    '''
```
I guess I need to add something to handle cases where `share_vocab` is True.

anderleich commented:
@Zenglinxiao 
> It seems the current update_vocab only supports adding new vocabulary to the previous one. What if we want to delete some?

Could you please give me some example of a use case? Why would we want to delete some tokens? Do they have an impact when not used during training? 

> Also, we could not change the order by increasing the frequency of some vocabulary in the current implementation.
For reference, in the OpenNMT-tf, they stored the mapping between the new vocab and old vocab which supports both add/delete/reorder.

Same. What impact has a different order by increasing the frequency? I guess it's just a mapping between tokens and intergers,  isn't it?

> PS: I personally prefer to make this a separate script creating an updated checkpoint rather than including it to train.py.

Does it make sense to update the checkpoint? I mean, we want to continue training from it so we only use it as a starting point. Maybe other use cases?

Zenglinxiao commented:
> Why would we want to delete some tokens?

After getting a trained model, we could find that some tokens in the vocabulary does not contribute much to the computation, by deleting them, we could reduce number of parameters.

> Same. What impact has a different order by increasing the frequency? I guess it's just a mapping between tokens and intergers, isn't it?

Yes, it's just a mapping. It won't impact the training, but could be a good feature to facilate analysis.

> Does it make sense to update the checkpoint? I mean, we want to continue training from it so we only use it as a starting point. Maybe other use cases?

Yes, we want to continue training using it as a starting point in most cases. In my opinion, saving it first rather than start right after update gives us the possibility to inspect the model.

The aforementioned suggestions are also based on the consideration of minimizing the difference with OpenNMT-tf.

anderleich commented:
@Zenglinxiao  I see most of the features you mentioned are more related to analysis rather than training. Currently I'm focusing on the training phase and the updating of the vocabulary by adding extra tokens. I would start by implementing this feature first which I find quite useful for my purposes. I open to consider improving and adding more functionalities in the future tough.
Do find this reasonable?

Zenglinxiao commented:
@anderleich Yes, for sure. Let's finish this and leave some feedback.

anderleich commented:
While I was correcting your comments I realized I was not taking into account `min_freq` and `max_size` paramenters for the vocabularies. For example, I we change `min_freq` for the new vocabulary, we should remove those tokens in the old vocabulary with a frequency less than `min_freq`. However, this arises some issues on how to delete those tokens from the old vocabulary, specially in the `itos` attribute of the vocabulary. Any ideas on how to solve this?

Zenglinxiao commented:
> One possible solution would be removing from `stoi` and `freqs` attributes and leaving `itos` as it is.

If we change one but not the other, this made the whole stuff unclear and inconsistent. Since we usually check `itos` to see the vocabulary list, this may be misleading. Regarding this issue, maybe it's still better to use the new vocab, and only remap the embeddings rather than update based on old fields. What's your opinion? @francoishernandez

francoishernandez commented:
I'm not sure it would be a good idea to change the `min_freq` while updating the vocab. This would create quite a mess.
As @Zenglinxiao  mentions, the best way would probably be to 1. build a new vocab and 2. remap the existing embeddings to this vocab. @anderleich would you feel comfortable implementing this?
I agree this would be a bit better in a standalone script. Though, once the logic is properly defined it won't be much of a problem to move it in its own script as a next step.

anderleich commented:
I've reimplemented the vocabulary updating procedure:

1. Run `build_vocab` as usual with the new datasets
2. Load checkpoint but override those fields with new fields with `prepare_fields_transforms`
3. In `load_base_model` check for each token in the new vocabulary if it exists in the checkpoint and if so update its embedding with theat from the checkpoint

This implementation is even simple and `min_freq` and `max_size` are handled as usual. I seems it works correctly but maybe there are some issues with `share_vocab` or when using custom embeddings (e.g GloVe).

anderleich commented:
I got the follwoing error on the optimizer. It seems there is still some information about the checkpoint vocabulary size.
```
Traceback (most recent call last):
  File "train.py", line 6, in <module>
    main()
  File "onmt/bin/train.py", line 172, in main
    train(opt)
  File "onmt/bin/train.py", line 157, in train
    train_process(opt, device_id=0)
  File "onmt/train_single.py", line 114, in main
    valid_steps=opt.valid_steps)
  File "onmt/trainer.py", line 244, in train
    report_stats)
  File "onmt/trainer.py", line 420, in _gradient_accumulation
    self.optim.step()
  File "onmt/utils/optimizers.py", line 367, in step
    self._optimizer.step()
  File "venv/lib/python3.6/site-packages/torch/autograd/grad_mode.py", line 15, in decorate_context
    return func(*args, **kwargs)
  File "venv/lib/python3.6/site-packages/torch/optim/adam.py", line 99, in step
    exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
**RuntimeError: The size of tensor a (29644) must match the size of tensor b (229) at non-singleton dimension 0**
```

francoishernandez commented:
I think it would be quite messy to try and update the optimizer here. (https://github.com/pytorch/pytorch/issues/1489#issuecomment-355301737 for instance)
Also, in some previous similar experiments I reset the optimizer and it trained fine.

anderleich commented:
`reset_optim="states"` seems to work just fine. However, I got some issues with training time on:
https://github.com/OpenNMT/OpenNMT-py/blob/94d96076b283a4b66c4c32e9db6ded13c511be56/onmt/trainer.py#L224
It seems to be in here:
```
for i, (batches, normalization) in enumerate(
                self._accum_batches(train_iter)):
```
It works, but too slow. Any guess? I think I haven't changed anything in there...

francoishernandez commented:
I don't understand. What does this performance question has to do with the vocab update?

anderleich commented:
I thought it was related to this update as it had not happened to me before. I've changed the dataset and it works as expected.
What I've seen is that for the previous dataset, it iterated the bucket four times for each batch. However, with the other dataset it iterates once for many batches, which is much faster

anderleich commented:
I think it is ready in terms of functionality. What should I do next?

francoishernandez commented:
As mentioned in [our earlier discussion](https://forum.opennmt.net/t/finetuning-with-the-new-version-opennmt-py-v2-0/4228/30?u=francoishernandez), can you add a functional test with a toy model / vocab, as well as an entry in the FAQ of the docs?
Thanks.

anderleich commented:
I've added a functional test and an entry in the FAQ

francoishernandez commented:
Thanks. Could you please also add your test in the [`pull_request_chk.sh`](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/tests/pull_request_chk.sh) script. We keep it in line with the CI config file to facilitate local testing.

As for the implementation, I think it's alright for a first version. Maybe just put the logic from `model_builder` in a separate function for better readability?
@Zenglinxiao can you have a look at the latest changes?

Zenglinxiao commented on onmt/model_builder.py:
```diff
@@ -232,6 +232,36 @@ def fix_key(s):
+            enc_emb_name = "encoder.embeddings.make_embedding.emb_luts.0.weight"
+            dec_emb_name = "decoder.embeddings.make_embedding.emb_luts.0.weight"
+
+            tgt_multifield = fields["tgt"]
+            checkpoint_tgt_multifield = checkpoint["vocab"]["tgt"]
```
Redundant

Zenglinxiao commented on onmt/model_builder.py:
```diff
@@ -232,6 +232,36 @@ def fix_key(s):
+            for field_name, emb_name in [("src", enc_emb_name), ("tgt", dec_emb_name)]:
+                multifield = fields[field_name]
+                checkpoint_multifield = checkpoint["vocab"][field_name]
+                for (name, field), (checkpoint_name, checkpoint_field) in zip(multifield, checkpoint_multifield):
```
multifield was designed for multiple source fields, there will be old models using the source features. As we've dropped that feature for now, maybe it's more appropriate to directly call `base_field` rather than iterate since the code doesn't handle updating other fields.

Zenglinxiao reviewed (commented):
The overall logic looks good to me. But lack of initialization for the new embedding. Imagine we have 70% vocab already exist in the old one, this only transfers the existing vocab's embedding into the new one while the 30% left not handle at all. We may want to first initialize the embedding, then transfer the weight.

anderleich commented:
I've moved the logic to a separate function and now embeddings are updated after parameter initalization.
I've also added a test in the `pull_request_chk.sh` script.

anderleich commented on onmt/model_builder.py:
```diff
@@ -232,9 +257,25 @@ def fix_key(s):
                                for k, v in checkpoint['model'].items()}
         # end of patch for backward compatibility
 
+        if model_opt.update_vocab:
+            # Remove old vocabulary associated embeddings
+            # Embedding layers
+            enc_emb_name = "encoder.embeddings.make_embedding.emb_luts.0.weight"
+            dec_emb_name = "decoder.embeddings.make_embedding.emb_luts.0.weight"
+            #Store before delete
+            checkpoint_backup = {"model": {}, "generator": {}}
+            checkpoint_backup["model"][enc_emb_name] = checkpoint["model"][enc_emb_name]
+            checkpoint_backup["model"][dec_emb_name] = checkpoint["model"][dec_emb_name]
+            checkpoint_backup["generator"]["0.weight"] = checkpoint["generator"]["0.weight"]
+            checkpoint_backup["generator"]["0.bias"] = checkpoint["generator"]["0.bias"]
+            checkpoint_backup["vocab"] = checkpoint["vocab"]
+            del checkpoint["model"][enc_emb_name], checkpoint["model"][dec_emb_name]
+            del checkpoint["generator"]["0.weight"], checkpoint["generator"]["0.bias"]
+
```
What would happen if checkpoint's embedding size is less than the models? `model.load_state_dict()` would only update those existing? That's why I removed the checkpoint's embedding layers

Zenglinxiao replied:
Oh, yes, you're right. I didn't notice that.

francoishernandez commented:
Not sure about this initialization thing. In the current state it seems we would lose all the checkpoint parameters except for the embeddings that we retrieve, instead of retaining the whole model and only add the necessary dimensions for the new vocab.

anderleich commented:
I've moved the parameter initilization to be before the checkpoint updating stuff. There will be 3 cases:

1. Normal training from scratch: only parameter initialization will be run
2. Continue from checkpoint: model will be updated with all the checkpoint's params
3. Update vocabulary: parameter initialization will be run, then vocabulary update, and finally the model will be updated with all the checkpoint's params except for those of the vocabulary

francoishernandez commented:
This looks better indeed.
Last thing before we can merge I think: it would probably be a good idea to add some check here:
https://github.com/OpenNMT/OpenNMT-py/blob/94d96076b283a4b66c4c32e9db6ded13c511be56/onmt/utils/parse.py#L259

E.g. `update_vocab` requires `train_from` not None, and maybe also warn that pretrained embeddings and update vocab are not compatible.

anderleich commented:
I've added some validation checks

francoishernandez commented:
Looks good! I'll merge, thanks!

anderleich commented:
Great!! 
I wondering which is the best option for `opt.reset_optim`, `states` or `all`? I mean should we initialize the learning rate, or keep the checkpoint's?

francoishernandez commented:
I think it might be a bit harsh to reset the learning rate in such cases. Would probably be worth experimenting to see the impact though.

anderleich commented:
I was wondering beacause maybe too small changes are made when learning from a new vocab and the system does not learn enough

francoishernandez commented:
Indeed in some cases it can be useful for instance to re-do a little bit of warmup when re-training/finetuning, but maybe not always. We should probably change the `assert` to allow both `reset_optim` `states` or `all` and people will decide which fits best to their use case.

francoishernandez commented:
Yes, please open a new PR and I'll merge. I'll probably bump the version as well so that the feature is exposed in the docs.

anderleich commented:
Ok.
Also this line causes model to update with checkpoints embedding on inference:
https://github.com/OpenNMT/OpenNMT-py/blob/cec3cc89d250bf2db905f8b1e14cb27fbefcb17e/onmt/train_single.py#L35

Any way to know wether it it training or inference?

francoishernandez commented:
You can probably disable here https://github.com/OpenNMT/OpenNMT-py/blob/cec3cc89d250bf2db905f8b1e14cb27fbefcb17e/onmt/model_builder.py#L81

which is the function used to load a model for inference.

anderleich commented:
I've opened a new PR

anderleich commented:
Hi, 
I have one question regarding the size of the vocabulary. I'm using BPE tokenization model trained on 32k operations. When processing the corpus it results in more tokens than the BPE operations (say 35.5k). Is this ok? Shouldn't we have a closed vocabulary of 32k tokens?
